capsules/src/segger_rtt.rs

```diff
@@ -91,77 +91,85 @@
 //! ```
 
 use core::cell::Cell;
-use kernel::common::cells::{OptionalCell, TakeCell};
+use core::marker::PhantomData;
+use kernel::common::cells::{OptionalCell, TakeCell, VolatileCell};
 use kernel::hil;
 use kernel::hil::time::Frequency;
 use kernel::hil::uart;
 use kernel::ReturnCode;
 
-/// Buffer for transmitting to the host.
-pub static mut UP_BUFFER: [u8; 1024] = [0; 1024];
+/// Suggested length for the up buffer to pass to the Segger RTT capsule.
+pub const DEFAULT_UP_BUFFER_LENGTH: usize = 1024;
 
-/// Buffer for receiving messages from the host.
-pub static mut DOWN_BUFFER: [u8; 32] = [0; 32];
+/// Suggested length for the down buffer to pass to the Segger RTT capsule.
+pub const DEFAULT_DOWN_BUFFER_LENGTH: usize = 32;
 
 /// This structure is defined by the segger RTT protocol. It must exist in
 /// memory in exactly this form so that the segger JTAG tool can find it in the
 /// chip's memory and read and write messages to the appropriate buffers.
 #[repr(C)]
-pub struct SeggerRttMemory {
-    id: [u8; 16],
-    number_up_buffers: u32,
-    number_down_buffers: u32,
-    up_buffer: SeggerRttBuffer,
-    down_buffer: SeggerRttBuffer,
+pub struct SeggerRttMemory<'a> {
+    id: VolatileCell<[u8; 16]>,
+    number_up_buffers: VolatileCell<u32>,
+    number_down_buffers: VolatileCell<u32>,
+    pub up_buffer: SeggerRttBuffer<'a>,
+    down_buffer: SeggerRttBuffer<'a>,
 }
 
 #[repr(C)]
-pub struct SeggerRttBuffer {
-    name: *const u8, // Pointer to the name of this channel. Must be a 4 byte thin pointer.
-    buffer: *const u8, // Pointer to the buffer for this channel.
-    length: u32,
-    write_position: u32,
-    read_position: u32,
-    flags: u32,
+pub struct SeggerRttBuffer<'a> {
+    name: VolatileCell<*const u8>, // Pointer to the name of this channel. Must be a 4 byte thin pointer.
+    pub buffer: VolatileCell<*const u8>, // Pointer to the buffer for this channel.
+    pub length: VolatileCell<u32>,
+    pub write_position: VolatileCell<u32>,
+    read_position: VolatileCell<u32>,
+    flags: VolatileCell<u32>,
+    _lifetime: PhantomData<&'a [u8]>,
 }
 
-impl SeggerRttMemory {
-    pub fn new(
+impl SeggerRttMemory<'a> {
+    pub fn new_raw(
         up_buffer_name: &'a [u8],
-        up_buffer: &'static mut [u8],
-        down_buffer_name: &'static [u8],
-        down_buffer: &'static mut [u8],
-    ) -> SeggerRttMemory {
+        up_buffer_ptr: *const u8,
+        up_buffer_len: usize,
+        down_buffer_name: &'a [u8],
+        down_buffer_ptr: *const u8,
+        down_buffer_len: usize,
+    ) -> SeggerRttMemory<'a> {
         SeggerRttMemory {
+            // TODO: only write this ID when the object is fully initialized, to avoid having
```

Comment:
Is this just a binary size optimization, or is there a correctness issue (i.e. where the debugger won't notice the buffer correctly) here?

Reply:
It's not entirely clear to me whether the debugger scans the flash memory. I haven't have any problems with the current implementation so far, but left this idea as a TODO comment.

Reply:
I haven't noticed any issue on the nRF52840-DK. I added a comment in case someone discovers a problem, but until then I think we can keep the current behavior of directly writing the magic value.